source/agora/common/Metadata.d

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+
+    Defines metadata which is stored to disk (e.g. peer info)
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Metadata;
+
+import agora.common.Data;
+import agora.common.Set;
+
+import vibe.core.log;
+import vibe.data.json;
+
+import std.array;
+import std.file;
+import std.format;
+import std.meta;
+import std.stdio;
+import std.typecons;
+
+/// Metadata stored to disk
+public class Metadata
+{
+    /// Known validators we want to try re-connecting to
+    public Set!Address validators;
+
+    /// Known listeners we may want to try re-connecting to
+    public Set!Address listeners;
+
+    /// Path to the data dir
+    private const string file_name;
+
+    /// Whether metadata was loaded on startup (and validators are not empty)
+    private bool loaded;
+
+    /// During unittesting we shouldn't do any file I/O
+    version(unittest) Appender!string test_appender;
+
+
+    /// Constructor
+    public this ( string data_dir )
+    {
+        this.file_name = format("%s/%s", data_dir, "metadata.dat");
+        version(unittest) this.test_appender = appender!string();
+    }
+
+    /// Dump the metadata to disk
+    public void dump ( )
+    {
+        version(unittest)
+        {
+            alias output = test_appender;
+        }
+        else
+        {
+            scope file = File(this.file_name, "w");
+            auto output = file.lockingTextWriter();
+        }
+
+        serializeToJson(output, tuple(validators, listeners));
+    }
+
+    /// Load any existing metadata from disk (if it exists),
+    /// and set 'this.loaded' to true if at least one validator was loaded
+    public void load ( )
```

Comment:
This class shouldn't do IO. Make it accept a string and take the IO out of it.

Reply:
Alright. I didn't think it was that important since it just stores IPs.